Preferences View Controller and UI

charliescheer opened the pull request:
### Fix
This is the first PR in the process of adding the Preferences window to Simplenote Mac (over all issue #1037, fixes #1038 )

In this PR I have added all of the UI for the preferences view and linked all of the controls into the window.

<img height=400 src="https://cdn-std.droplr.net/files/acc_593859/KCRls9"/>

Note: one thing that is not done here is the button color for the Delete Account  button.  Still working on getting that going.

### Test
1. Test opening the preferences window by going to Simplenote > Preferences, hitting Command + , and clicking on the Settings button in the tag list
2. Make sure that you can only open one preferences view at a time (shouldn't be possible to have multiple windows of preferences appear)
3. Check and see if your account email appears in the top
4. Test log out button
5. log in with a testing account, test delete account button
6. change the sort order of your notes and make sure the note list updates
7. change the line length setting from full to narrow and back
8. select the condensed note list and deselect it
9. select the sort tags alphabetically and deselect it
10. try the different theme settings
11. use the font slider to change the font size
12. test to make sure these setting changes persist when you close the preferences window and when you restart the app
13. click on the "About Analytics and Privacy" link to be taken to the a8c privacy page

### Review
> Only one developer and one designer are required to review these changes, but anyone can perform the review.

### Release

> `RELEASE-NOTES.txt` was updated in a4948377 with:
> 
> > Added preferences menu to centralize app options and customizations #1042

charliescheer committed:
Added UI elements for PreferencesViewController

charliescheer committed:
Added outlet connections to the Preferences View Controller

charliescheer committed:
Added outlets for all controls on Preferences VC

charliescheer committed:
Added logic to display preferences from menu and shortcut

charliescheer committed:
Set outlets and actions to private in privacy vc

charliescheer committed:
Added logout functions to PreferencesViewController

charliescheer committed:
added account deletion to PreferencesViewController

charliescheer committed:
Added sort mode logic to PreferencesViewController

charliescheer committed:
Added Note Line Length to preferencesVC

charliescheer committed:
Fixed a problem with the selected sortmode  persisting in preferences

charliescheer committed:
Added condensed list display to the Preferences View Controller

charliescheer committed:
Fixed condensed note display in Preferences View Controller

charliescheer committed:
Added sort tags alphabetically to Preferences View Controller

charliescheer committed:
Added theme setting to Preferences View Controller

charliescheer committed:
Set scrolling limits on text size

charliescheer committed:
Added share analytics setting to Preferences View Controller

charliescheer committed:
Added about page button to preferences View controller

charliescheer committed:
Added font size settings to Preferences View Controller

charliescheer committed:
Fixed issue where multiple preferences windows could be opened

charliescheer committed:
Added tracks event for opening preferences

charliescheer committed:
Changed available font selection sizes on preferences VC

charliescheer committed:
Removed direct calls to user defaults for font size in favor of options

charliescheer committed:
Added settings button to tag list view controller

charliescheer committed:
Fixed an issue with the height of the preferences view height

charliescheer committed:
Fixed an issue with the Preferences VC height

charliescheer committed:
Added red color for delete account button

charliescheer committed:
Removed connected outlets that were not being used in preferences VC

charliescheer committed:
Removed About button from preferences

charliescheer committed:
Moved setup of preferences shortcut to storyboard

charliescheer committed:
Resetup the preferences storyboard to use stackview and custom views

charliescheer committed:
Removed IBAction for displaying the about window

charliescheer committed:
Added outlets for all labels to PreferencesVC

charliescheer committed:
Changed label outlets to private

charliescheer committed:
Removed stored options instance in Preferences VC

charliescheer committed:
Removed stored about WC from preferences VC

charliescheer committed:
Added link to privacy policy in preferences view controller

charliescheer committed:
Fixed some sizing issues

charliescheer committed:
Localized strings in Preferences View Controller

charliescheer committed:
Set correct colors for dark and light mode in preferences vc

charliescheer committed:
Set correct colors for button text on preferences vc

charliescheer committed:
Updated release notes adding preferences view controller

charliescheer committed:
Updated styling for delete account button

charliescheer committed:
Added dividers between different sections in PreferncesViewController

charliescheer committed:
Changed menu font size options to step by 5 instead of one

jleandroperez reviewed (commented):
@charliescheer sending you few UX notes, getting to the actual code in about a second!! Thank you!!

### Preferences Window
I've noticed the `Maximize` button is enabled. Clicking on it toggles, obviously!!, the Preferences UI into a fullscreen window.

<img width="200" src="https://user-images.githubusercontent.com/1195260/135340903-2bf5f91f-af88-43c4-8a15-8362139ea809.jpg">

### Theme Switch
1. Open the Preferences UI
2. Switch over to a different theme

Notice the Preferences UI doesn't reflect the newly selected Theme.

### Preferences Sync
1. Open the Preferences UI
2. While the Preferences UI is onscreen, update any of the settings, but from the App's Menu

Notice that the Preferences UI falls out of sync with the actual underlying setting.

### Menu
@SylvesterWilmott can we drop `Log Out` and `Delete Account` from the `Simplenote` macOS App Menu? (feels like it'd be okay to have those two _only_ in the Preferences UI).

charliescheer committed:
Updated preferences window to respond to theme changes and not maximize

charliescheer commented:
@jleandroperez Thanks for the feedback.

>I've noticed the Maximize button is enabled. Clicking on it toggles, obviously!!, the Preferences UI into a fullscreen window.

I've removed the minimize and maximize buttons

>Notice the Preferences UI doesn't reflect the newly selected Theme.

Oh nice catch! I fixed that.


We will be removing most of the menu links in another PR, so falling out of sync shouldn't be an issue.  We are keeping the font size changes, so will need to do something about making sure those are up to date, but the rest shouldn't be an issue

jleandroperez commented on Simplenote/SimplenoteConstants.swift:
```diff
@@ -31,4 +31,11 @@ class SimplenoteConstants: NSObject {
+    static let NoteFontSizeMinimum = 10
+    static let NoteFontSizeNormal = 15
+    static let NoteFontSizeMaximum = 30
+    static let NoteFontSizeStep = 5
```
Since all of these are prefixed as `NoteFontSize`, it would probably be a good idea to move it to its own enum / struct.

How about... `FontSettings`?  (Oh! another note, constants should not be capitalized in Swift) 😄

charliescheer replied:
>another note, constants should not be capitalized in Swift)

A shoot that is a copy/paste error on my part.  Will move this to an enum somewhere

jleandroperez commented on External/Notepad/Theme.swift:
```diff
@@ -17,7 +17,7 @@ class Theme: NSObject {
     /// Default Font Size
     ///
-    private static let defaultFontSize = CGFloat(14)
+    private static let defaultFontSize = CGFloat(SimplenoteConstants.NoteFontSizeNormal)
```
It'd probably be a good idea to drop this constant, since it's a constant of a constant.

Plus: noted a bit below, since there are a bunch of constants named `NoteFontSize`, we could probably group them into their own namespace. Perhaps `FontSettings.defaultFontSize` ? 

Or... `FontConstants`, but `FontSettings` somehow feels cooler?

jleandroperez commented on External/Notepad/Theme.swift:
```diff
@@ -96,11 +96,11 @@ private extension Theme {
 
     private static var fontSize: CGFloat {
-        var fontSize = CGFloat(UserDefaults.standard.integer(forKey: "kFontSizePreferencesKey"))
+        var fontSize = CGFloat(Options.shared.fontSize)
```
Q: Can we have the `fontSize` be a `CGFloat` directly, rather than converting here?

jleandroperez commented on External/Notepad/Theme.swift:
```diff
@@ -96,11 +96,11 @@ private extension Theme {
 private extension Theme {
 
     private static var fontSize: CGFloat {
```
Perhaps this entire method could be taken away (and replaced by, directly, using fontSize, granted that we patch it up to return the default Size if nothing is set)

jleandroperez commented on Simplenote/Int+Extensions.swift:
```diff
@@ -0,0 +1,7 @@
+extension Int {
+    func roundToNearestFive() -> Int {
+        return 5 * Int((Double(self) / 5.0).rounded())
+    }
```
Since this API is *not* generic (it's very, very specific to just one use case), perhaps it could be better off as a function in `FontSettings` ? (or elsewhere?)

jleandroperez replied:
Second idea would be to parametrize. `roundToNearest(multiple: Double)` (or something like that, but idea is to drop the constant)

jleandroperez commented on Simplenote/NSImageName+Simplenote.swift:
```diff
@@ -38,6 +38,10 @@ extension NSImage.Name {
 
+    /// Tags List: Settings
+    ///
+    static let settings = NSImage.Name("button_settings")
```
Really Nipticky: constants in this file are sorted by name 😂 

Told you. Seriously nipticky!!

jleandroperez commented on Simplenote/NoteEditorViewController+Swift.swift:
```diff
@@ -926,6 +926,18 @@ extension NoteEditorViewController {
     }
+
+    @objc
+    func getRoundedFontSize() -> Int {
```
Can we pleaaaase drop the`get`? (since it's not swift-y). Some of the old code, though, has that, buuut we should 🔥

Plus: perhaps `FontSettings` would be a more appropriate location?

jleandroperez commented on Simplenote/NoteEditorViewController+Swift.swift:
```diff
@@ -926,6 +926,18 @@ extension NoteEditorViewController {
+    @objc
+    func getRoundedFontSize() -> Int {
+        var currentFontSize = Options.shared.fontSize
+        let fontSizeIsNotValid = currentFontSize % SimplenoteConstants.NoteFontSizeStep != .zero
```
Mind verifying if we **do** need this? it's possible that calling the rounding method, if it's already rounded, won't do anything.

jleandroperez commented on Simplenote/PreferencesViewController.swift:
```diff
@@ -0,0 +1,314 @@
+    @IBOutlet var accountSectionBackground: BackgroundView!
+    @IBOutlet var layoutSectionBackground: BackgroundView!
+    @IBOutlet var themeSectionBackground: BackgroundView!
+    @IBOutlet var textSectionBackground: BackgroundView!
```
`private`

jleandroperez commented on Simplenote/PreferencesViewController.swift:
```diff
@@ -0,0 +1,314 @@
+import Cocoa
+
+class PreferencesViewController: NSViewController {
+    let simperium = SimplenoteAppDelegate.shared().simperium
```
This one shouldn't be a constant (perhaps a dynamic property?).

We're not doing it, but in theory, the `Simperium` instance could be swapped, and we could be left there with an old reference.

Plus: `private`

jleandroperez commented on Simplenote/PreferencesViewController.swift:
```diff
@@ -0,0 +1,314 @@
+
+        shareAnalyticsCheckbox.state = Options.shared.analyticsEnabled ? .on: .off
+
+    }
```
I've noticed some components have an `update` API, some don't. Would it be possible to make it all standard?

(Plus: extra newline!). TY!!

jleandroperez commented on Simplenote/PreferencesViewController.swift:
```diff
@@ -0,0 +1,314 @@
+            return item
+        }
+        
+        menuItems.forEach({ sortOrderPopUp.menu?.addItem($0) })
```
Both `SortMode.allCases` are being iterated twice (by `map`, and then in the `forEach`).

Plus, perhaps we should unwrap `sortOrderPopUp` before doing anything (and assert on that), otherwise this API will fail to do anything, but we wouldn't notice.

jleandroperez commented on Simplenote/PreferencesViewController.swift:
```diff
@@ -0,0 +1,314 @@
+            return item
+        }
+
+        menuItems.forEach({ themePopUp.menu?.addItem($0) })
```
Same note as above

jleandroperez commented on Simplenote/SimplenoteAppDelegate+Swift.swift:
```diff
@@ -160,6 +160,24 @@ extension SimplenoteAppDelegate {
+        SPTracker.trackPreferencesWasOpened()
+
+        if let preferencesWindow = preferencesWindowController?.window {
+            if preferencesWindow.isVisible {
```
You can collapse these two `if` statements into a single one:

```
if preferencesWindowController?.window?.isVisible == true
```

jleandroperez commented on Simplenote/SimplenoteAppDelegate+Swift.swift:
```diff
@@ -602,4 +620,6 @@ extension SimplenoteAppDelegate {
 //
 private struct Constants {
     static let noteEditorMetadataCacheFilename = ".editor-metadata-cache"
+    static let preferenceStoryboardIdentifier = "Preferences"
```
No need for this constant. Check `NSStoryboard+Simplenote`

jleandroperez commented on Simplenote/SimplenoteAppDelegate+Swift.swift:
```diff
@@ -602,4 +620,6 @@ extension SimplenoteAppDelegate {
 private struct Constants {
     static let noteEditorMetadataCacheFilename = ".editor-metadata-cache"
+    static let preferenceStoryboardIdentifier = "Preferences"
+    static let preferencesWindowControllerIdentifier = "PreferencesWindowController"
```
No need for this constant either. Check on the `instantiateWindowController` NSStoryboard extension 😄

jleandroperez commented on Simplenote/SimplenoteAppDelegate.h:
```diff
@@ -49,6 +49,8 @@ NS_ASSUME_NONNULL_BEGIN
 
 @property (strong, nonatomic) NoteEditorMetadataCache                       *noteEditorMetadataCache;
 
+@property (nullable, strong, nonatomic) NSWindowController                            *preferencesWindowController;
```
Nipticky: different spacing with regards of the other properties (yes, I know, we should probably not do this!!)

charliescheer committed:
Fixed an issue where the font size display sync  if changed in menu

charliescheer commented:
I've added another commit that fixes the font size getting out of sync in the preferences view controller

jleandroperez commented on Simplenote/TagListViewController+Swift.swift:
```diff
@@ -218,6 +220,12 @@ extension TagListViewController: NSTableViewDataSource, SPTableViewDelegate {
+           selectedRow == .settings {
+            SimplenoteAppDelegate.shared().preferencesWasPressed(self)
+            return
+        }
```
There's a UX problem with having `Settings` in the Tags list:

1. Click on the `Settings` row
2. Preferences UI shows up
3. Close the preferences window

- The Tags list remains highlighting `Settings`, but there are no more settings onscreen.
- While `Settings` is highlighted, it's not clear what `Notes` are being filtered out.

Since it creates a few nipticky UX problems, @SylvesterWilmott do you think we could drop the `Settings` row from the Tags List?

jleandroperez reviewed (commented):
Thank you Charlieeee!!! apologies in advance about the annoying notes!!

charliescheer committed:
Refactored Options.fontSettings to store a CGFloat instead of Int

charliescheer committed:
Moved font calculations and sizes to FontSettings class

charliescheer committed:
Removed int+extensions file and method

charliescheer committed:
moved settings image name to alphabetical order

charliescheer committed:
Updated property declarations for preferences vc

charliescheer committed:
Updated refreshing values in preferences to all have a method

charliescheer committed:
Refactored the creation of the sort mode and theme popup menu fields

charliescheer committed:
Simplified the guard against multiple preferences view controllers

charliescheer committed:
Refactored the instantiation of the preferences window
